transport/transport.test.ts

```diff
@@ -536,6 +538,165 @@ describe.each(testMatrix())(
+      await waitFor(() => expect(serverTransport.connections.size).toEqual(1));
+
+      const oldClientSession = serverTransport.sessions.get('client');
+      const oldServerSession = clientTransport.sessions.get('SERVER');
+      expect(oldClientSession).toMatchObject({
+        id: oldServerSession?.advertisedSessionId,
+      });
+      expect(oldServerSession).toMatchObject({
+        id: oldClientSession?.advertisedSessionId,
+      });
+
+      clientTransport.reconnectOnConnectionDrop = false;
+      clientTransport.connections.forEach((conn) => conn.close());
+      await waitFor(() => expect(clientTransport.connections.size).toEqual(0));
+      await waitFor(() => expect(serverTransport.connections.size).toEqual(0));
+
+      clientTransport.reconnectOnConnectionDrop = true;
+      await clientTransport.connect('SERVER');
+      await waitFor(() => expect(clientTransport.connections.size).toEqual(1));
+      await waitFor(() => expect(serverTransport.connections.size).toEqual(1));
+
+      // expect new sessions to have been created
+      const newClientSession = serverTransport.sessions.get('client');
+      const newServerSession = clientTransport.sessions.get('SERVER');
+      expect(newClientSession).not.toBe(oldClientSession);
+      expect(newServerSession).not.toBe(oldServerSession);
+
+      await testFinishesCleanly({
+        clientTransports: [clientTransport],
+        serverTransport,
+      });
+    });
+  },
+);
+
+describe.each(testMatrix())(
+  'transport handshake grace period tests ($transport.name transport, $codec.name codec)',
+  async ({ transport, codec }) => {
+    const opts: TransportOptions = {
+      codec: codec.codec,
+      heartbeatIntervalMs: 1_000,
+      heartbeatsUntilDead: 2,
+      sessionDisconnectGraceMs: 10_000,
+      // setting session grace to be higher so that only handshake grace passes
+      handshakeTimeoutMs: 500,
+    };
+    const clientOpts: ClientTransportOptions = {
+      ...opts,
+      attemptBudgetCapacity: 1,
+    };
+    let testHelpers: TestSetupHelpers;
+    let getClientTransport: TestSetupHelpers['getClientTransport'];
+    let getServerTransport: TestSetupHelpers['getServerTransport'];
+    const { addPostTestCleanup, postTestCleanup } = createPostTestCleanups();
+    beforeEach(async () => {
+      testHelpers = await transport.setup({ client: clientOpts, server: opts });
+      getClientTransport = testHelpers.getClientTransport;
+      getServerTransport = testHelpers.getServerTransport;
+      return async () => {
+        await postTestCleanup();
+        await testHelpers.cleanup();
+      };
+    });
+
+    test('handshake grace period of 0 should lead to closed connections', async () => {
+      const schema = Type.Unknown();
+      const get = vi.fn(async () => ({}));
+
+      const parse = vi.fn(async () => {
+        const promise = new Promise(() => {
+          // noop we never want this to return
+        });
+        return promise;
+      });
```

Comment:
```suggestion
      const parse = vi.fn(() => {
        const promise = new Promise(() => {
          // noop we never want this to return
        });
        return promise;
      });
```

don't need async there